OsmAnd/src/net/osmand/plus/views/layers/PointLocationLayer.java

```diff
@@ -69,6 +86,186 @@
+	private static final int MARKER_ID_MY_LOCATION = 1;
+	private static final int MARKER_ID_NAVIGATION = 2;
+	private static final int MARKER_ID_OUTDATED_LOCATION = 3;
+	private MapMarkersCollection markersCollection;
+	private MapMarker myLocationMarker;
+	private MapMarker navigationMarker;
+	private MapMarker outdatedLocationMarker;
+	private SWIGTYPE_p_void onSurfaceIconKey = null;
+	private SWIGTYPE_p_void onSurfaceHeadingIconKey = null;
+	private boolean dirty = true;
+
+	private enum State {
+		MarkerStateStay,
+		MarkerStateMove,
+		MarkerStateOutdatedLocation
+	}
+	private State currentMarkerState = State.MarkerStateOutdatedLocation;
+
+	private void setMarkerState(State markerState) {
+		if (currentMarkerState == markerState) {
+			return;
+		}
+		currentMarkerState = markerState;
+		updateMarkerState();
+	}
+
+	private byte[] getBitmapAsByteArray(@NonNull Bitmap bitmap) {
+		int size = bitmap.getRowBytes() * bitmap.getHeight();
+		ByteBuffer byteBuffer = ByteBuffer.allocate(size);
+		bitmap.copyPixelsToBuffer(byteBuffer);
+		return byteBuffer.array();
+	}
+
+	private MapMarker recreateMarker(MapMarker oldMarker, LayerDrawable icon, int id, int baseColor) {
+		if (view == null) {
+			return null;
+		}
+
+		final MapRendererView mapRenderer = view.getMapRenderer();
+		if (mapRenderer == null) {
+			return null;
+		}
+
+		boolean newCollection = false;
+		if (markersCollection == null) {
+			markersCollection = new MapMarkersCollection();
+			newCollection = true;
+		} else if (oldMarker != null) {
+			oldMarker.setIsHidden(true);
+			oldMarker.setIsAccuracyCircleVisible(false);
+			markersCollection.removeMarker(oldMarker);
+		}
+
+		MapMarkerBuilder myLocMarkerBuilder = new MapMarkerBuilder();
+		myLocMarkerBuilder.setMarkerId(id);
+		myLocMarkerBuilder.setIsAccuracyCircleSupported(true);
+		myLocMarkerBuilder.setAccuracyCircleBaseColor(new FColorRGB((baseColor >> 16 & 0xff)/255.0f,
+				((baseColor >> 8) & 0xff)/255.0f,
+				((baseColor) & 0xff)/255.0f));
+		myLocMarkerBuilder.setPinIconVerticalAlignment(MapMarker.PinIconVerticalAlignment.CenterVertical);
+		myLocMarkerBuilder.setPinIconHorisontalAlignment(MapMarker.PinIconHorisontalAlignment.CenterHorizontal);
+		myLocMarkerBuilder.setIsHidden(true);
+
+		Bitmap markerBitmap = AndroidUtils.createScaledBitmap(icon, icon.getIntrinsicWidth(), icon.getIntrinsicHeight());
+		if (markerBitmap != null) {
+			SWIGTYPE_p_sk_spT_SkImage_const_t swigImg = SwigUtilities.createSkImageARGB888With(
+					markerBitmap.getWidth(), markerBitmap.getHeight(), getBitmapAsByteArray(markerBitmap));
+			onSurfaceIconKey = SwigUtilities.getOnSurfaceIconKey(1);
+			myLocMarkerBuilder.addOnMapSurfaceIcon(onSurfaceIconKey, swigImg);
+		}
+
+		if (id != MARKER_ID_OUTDATED_LOCATION) {
+			Bitmap headingBitmap = getScaledBitmapWithTint(headingIconId, getTextScale(), baseColor);
+			if (headingBitmap != null) {
+				SWIGTYPE_p_sk_spT_SkImage_const_t swigImg = SwigUtilities.createSkImageARGB888With(
```

Comment:
Would be better to create NativeUtilities and make method like createSkImageFromBitmap